### PR TITLE
[Core] Support fp32 lm_head for generation models via head_dtype (RFC #48305 §3.6)

### PR DESCRIPTION
**Pull request by KKothuri**
## Purpose

RL training-inference consistency (RFC #48305 §3.6, "Dtype Replay") requires
vLLM's rollout logits to match a trainer that computes the lm_head in fp32.

vLLM already has `ModelConfig.head_dtype` (merged for pooling models in #23810),
but it was **restricted to pooling models** — generation models silently fell
back to the model dtype. This PR **extends `head_dtype` to generation models**
(the goal of the stalled #24567; fixes #19925): it lifts the pooling-only
restriction and honors `head_dtype` in `LogitsProcessor._get_logits`.

Enable with `--hf-overrides '{"head_dtype": "float32"}'`
(or `LLM(hf_overrides={"head_dtype": "float32"})`). Restricted to an unquantized
lm_head. Default behavior is unchanged.

## Why this isn't a duplicate

The related non-contiguous-logits NaN in the top-k/top-p kernel was already
fixed by #42739; that did not make an fp32 generation head usable. Current
ecosystem workarounds: prime-rl monkeypatches vLLM's `LogitsProcessor`
(`monkey_patch_fp32_lm_head`), Flash-RL patches similarly, sglang has a native
`enable_fp32_lm_head`, and slime inherits it via sglang; SkyRL uses its own
inference stack. This provides the first-class vLLM path through the **existing**
`head_dtype` knob rather than a new config, continuing #23810/#24567.

## Test

- Unit (`tests/v1/sample/test_head_dtype.py`): fp32 projection numerics, no-op
  when `head_dtype == model dtype`, and the quantized-lm_head guard — pass.
- E2E on the **v2 model runner**: `head_dtype="float32"` + `processed_logprobs`,
  finite logprobs, no NaN.
- gsm8k (Qwen2.5-0.5B-Instruct, 5-shot, full 1319, greedy, flexible-extract):
  bf16 head **0.339** / fp32 head **0.350** (±0.013). Within noise — the gap
  flips sign vs a limit=200 run (there fp32 was 0.015 lower), confirming no
  accuracy effect. The feature targets logprob fidelity for RL, not greedy
  accuracy.

## Notes

- The CUDA lm_head projection uses `torch.mm(..., out_dtype=torch.float32)`
  (native bf16×bf16→fp32, the same technique as prime-rl), which avoids
  materializing a per-step fp32 copy of the lm_head weight. Non-CUDA backends
  fall back to casting both operands (`out_dtype` mm is CUDA-only).
- `--enable-lora` combined with a non-default `head_dtype` is rejected for now:
  the LoRA logits path (`LogitsProcessorWithLoRA`) bypasses this projection and
  would otherwise silently emit model-dtype logits.
- AI assistance (Claude Code) was used; I reviewed every line.

### CODE REVIEW

**Comment by aoshen02** [comment]
@codex review

File: tests/v1/sample/test_logits_dtype.py [hunk 1]
```diff
@@ -0,0 +1,53 @@
+# SPDX-License-Identifier: Apache-2.0
+# SPDX-FileCopyrightText: Copyright contributors to the vLLM project
+"""Tests for the fp32 lm_head projection (``logits_dtype='float32'``).
+
+Casting the lm_head to fp32 lets rollout logits match a trainer that computes
+logits in fp32, which is required for RL training-inference consistency.
```
Comment (aoshen02) [inline tests/v1/sample/test_logits_dtype.py]: Please also add test to model runner v2?

**Comment by noooop** [comment]
Thanks for your help with implementing this feature.

**Comment by aoshen02** [comment]
1. I think prime rl practice make a lot of sense, we should adopt.
2. I think 0.015 is definitely small noisy, but could you please run it one more time, want to see the result.

**Comment by KKothuri** [comment]
> 1. I think prime rl practice make a lot of sense, we should adopt.
> 2. I think 0.015 is definitely small noisy, but could you please run it one more time, want to see the result.

I implemented the prime_rl approach for CUDA envs. Updated the description.

Ran the full 1319-problem set: bf16 0.339 / fp32 0.350 flexible-extract (±0.013).

**Comment by aoshen02** [comment]
cc@SumanthRH

**Comment by aoshen02** [comment]
@codex review

**Comment by aoshen02** [comment]
@claude review

**Comment by aoshen02** [comment]
@codex review

**Comment by SumanthRH** [comment]
Thanks for the PR @KKothuri, looks good @aoshen02

**Review by noooop — APPROVED** [review]

**Comment by aoshen02** [comment]
Could you check this path? I think dtype=fp32 won't work:
use_local_argmax_reduction=True
    -> model.get_top_tokens(...)
    -> LogitsProcessor.get_top_tokens(...)
    -> lm_head.quant_method.apply(...)

**Comment by aoshen02** [comment]
Could you do a follow-up pr about lora path?

File: tests/models/language/generation_ppl_test/test_gpt.py [hunk 2]
```diff
@@ -11,4 +11,14 @@
...
+    fp32_ppl = wikitext_ppl_test(
+        hf_runner,
+        vllm_runner,
+        model_info,
+        vllm_extra_kwargs={"hf_overrides": {"head_dtype": "float32"}},
+    )
+
+    differ = ((fp32_ppl - bf16_ppl) / bf16_ppl) * 100
+    print("fp32 head difference (%):", differ)
+    assert differ < 0
```
Comment (noooop) [inline tests/models/language/generation_ppl_test/test_gpt.py:24]: https://buildkite.com/vllm/ci/builds/77747/canvas?jid=019f59bf-cb0b-4950-80cc-e95dbd99e668&tab=output

Model: openai-community/gpt2-large
VLLM: dtype:torch.bfloat16 head_dtype:torch.bfloat16 19.447505950927734
Transformers: Constant 19.457056045532227
Difference (%): -0.04908293722412904

Model: openai-community/gpt2-large
VLLM: dtype:torch.bfloat16 head_dtype:torch.float32 19.44658851623535
Transformers: Constant 19.457056045532227
Difference (%): -0.05379811453685244
fp32 head difference (%): -0.004717492796751383

The smaller the PPL test value, the better, which indicates that head_dtype: torch.float32 is working (Using V2 Model Runner).
